MMS-98 - Artifact Storage

BA-MBSE opened the pull request.

HuiJun commented on mms-ent/repo-amp/src/main/java/gov/nasa/jpl/view_repo/db/ElasticHelper.java:
```diff
@@ -2,13 +2,10 @@
+import java.util.*;
 
 import com.google.gson.JsonElement;
+import io.searchbox.core.*;
```
Explicitly import

EDKarlsson replied:
Fixed

HuiJun commented on mms-ent/repo-amp/src/main/java/gov/nasa/jpl/view_repo/db/ElasticHelper.java:
```diff
@@ -464,4 +455,62 @@ private static JSONObject removeWrapper(JSONObject jsonObject) {
+        query.put("query", new JSONObject().put("term", new JSONObject().put(field, id)));
+
+        // Verbose statement to make sure it uses the correct delete by query class from searchbox.
+        DeleteByQuery deleteByQuery = new io.searchbox.core.DeleteByQuery.Builder(query.toString()).addIndex(elementIndex).build();
```
Clean up

HuiJun commented on mms-ent/repo-amp/src/main/java/gov/nasa/jpl/view_repo/db/PostgresHelper.java:
```diff
@@ -2,11 +2,7 @@
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.sql.Savepoint;
+import java.sql.*;
```
Explicit import

HuiJun commented on mms-ent/repo-amp/src/main/java/gov/nasa/jpl/view_repo/db/PostgresHelper.java:
```diff
@@ -2100,4 +2096,46 @@ public boolean refExists(String refId) {
+     */
+    public void deleteProjectFromProjectsTable(String projectId){
+        String query = "DELETE FROM projects WHERE projectid = \'" + projectId +"\'";
+        System.out.println("Delete ProjectId Query "  + query);
```
Use prepared statement here

HuiJun replied:
Also remove println

EDKarlsson replied:
I tried making this a prepared statement but it throws a null pointer exception on the `PreparedStatment query =...` line.
```
PreparedStatement query = this.conn.prepareStatement("DELETE FROM projects WHERE projectid = ?");
query.setString(1, projectId);
query.executeUpdate();
```
2017-06-29 08:47:58,221  ERROR [view_repo.webscripts.ProjectDelete] [http-bio-8080-exec-1] java.lang.NullPointerException
	at gov.nasa.jpl.view_repo.db.PostgresHelper.deleteProjectFromProjectsTable(PostgresHelper.java:2137)
	at gov.nasa.jpl.view_repo.webscripts.ProjectDelete.deleteProjectFromProjectsTable(ProjectDelete.java:177)
	at gov.nasa.jpl.view_repo.webscripts.ProjectDelete.executeImplImpl(ProjectDelete.java:72)
	at gov.nasa.jpl.view_repo.webscripts.ProjectDelete.executeImpl(ProjectDelete.java:47)
	at gov.nasa.jpl.view_repo.webscripts.DeclarativeJavaWebScript.execute(DeclarativeJavaWebScript.java:103)
   ...
```

HuiJun commented on mms-ent/repo-amp/src/main/java/gov/nasa/jpl/view_repo/db/PostgresPool.java:
```diff
@@ -69,4 +69,18 @@ public static Connection getStandaloneConnection(String host, String name) throw
+    public static void removeConnection(String host, String name) {
+        String connectString = host + name;
+
+        System.out.println("Removing connection " + connectString);
```
Remove

BA-MBSE committed:
Adding artifact interfaces

BA-MBSE committed:
Continuing to flesh out Artifact handling

BA-MBSE committed:
Bug fix

BA-MBSE committed:
Added versioning to InMemoryArtifactStorage

BA-MBSE committed:
Adding tests for artifacts.  All work except the last one - still need to deal with validating artifacts coming in through the elements endpoints.

BA-MBSE committed:
Merge branch 'develop' into feature/MMS-98

BA-MBSE committed:
Continuing work on artifacts

BA-MBSE committed:
Added PreCommitSubscriber infrastructure to allow a generic way to do pre-commit work
Disallow changes to artifacts in Elements if the changes don't come through the ArtifactController
updated postman tests

BA-MBSE committed:
Added delete test

BA-MBSE committed:
Added interceptor to allow different endpoint behavior for the get element endpoint based on the Accept header (plug-and-play if artifacts module is included)

BA-MBSE committed:
Added more artifact tests, fixed a few bugs

BA-MBSE committed:
Changed PreCommitService to allow optional autowiring (if no precommit subscribers exist)

dlamoris commented on artifacts/src/main/java/org/openmbee/sdvc/artifacts/controllers/ArtifactController.java:
```diff
@@ -0,0 +1,86 @@
+        params.put(ArtifactConstants.EXTENSION_PARAM, extension);
+        ArtifactResponse artifact = artifactService.get(projectId, refId, elementId, params);
+
+        if(artifact == null) {
```
instead of checking for null here, would be better if the artifact service call just throws an exception so it'll be caught by the controller advice

dlamoris commented on crud/src/main/java/org/openmbee/sdvc/crud/controllers/elements/ElementsController.java:
```diff
@@ -71,6 +71,8 @@ public ElementsResponse createOrUpdateElements(
         throw new BadRequestException(response.addMessage("Empty"));
     }
 
+
```
comment is for the /elementId getElement - if you add produces = MediaType.APPLICATION_JSON_VALUE, it won't throw an error if there's a same mapping with no produces in artifacts controller

BA-MBSE replied:
The story called for the elements controller to handle JSON and empty Accept headers. There's nothing to handle empty Accept headers if produces is defined and the artifacts module is not included in a build.

BA-MBSE replied:
handled via content negotiation

dlamoris commented on example/src/main/java/org/openmbee/sdvc/example/config/ExampleSecurityConfig.java:
```diff
@@ -96,4 +103,11 @@ private CorsFilter corsFilter() {
+        if(interceptors != null) {
+            interceptors.forEach(v -> v.register(registry));
+        }
+    }
```
the interceptor shouldn't be needed if content negotiation is used - https://www.baeldung.com/spring-mvc-content-negotiation-json-xml
something like this
```
@Override
    public void configureContentNegotiation(ContentNegotiationConfigurer configurer) {
        configurer.favorPathExtension(false).
            favorParameter(false).
            ignoreAcceptHeader(false).
            useJaf(false).
            defaultContentType(MediaType.APPLICATION_JSON);
    }
```
this sends requests that don't have accept defined to default to json, along with adding `produces` to the elements get controller method, there can be another controller method for artifacts with the same url mapping (but no produces) so it can handle anything else with accept

BA-MBSE replied:
Ok, I'll look into this

BA-MBSE replied:
Reimplemented using this method.  Much cleaner now, thanks!

dlamoris commented on artifacts/src/main/java/org/openmbee/sdvc/artifacts/service/DefaultArtifactService.java:
```diff
@@ -0,0 +1,211 @@
+    public ArtifactResponse get(String projectId, String refId, String id, Map<String, String> params) {
+        NodeService nodeService = getNodeService(projectId);
+        ElementJson elementJson = getElement(nodeService, projectId, refId, id, params);
+        if(elementJson == null){
```
prefer throwing not found exceptions than returning null

BA-MBSE replied:
now throwing instead

dlamoris commented on artifacts/src/main/java/org/openmbee/sdvc/artifacts/service/DefaultArtifactService.java:
```diff
@@ -0,0 +1,211 @@
+        elementJson = attachOrUpdateArtifact(elementJson, artifactLocation, fileExtension, mimeType, "internal");
+        ElementsRequest elementsRequest = new ElementsRequest();
+        elementsRequest.setElements(Arrays.asList(elementJson));
+        nodeService.createOrUpdate(projectId, refId, elementsRequest, params, user);
```
can just return the response from nodeService, there's no guarantee the elementJson is updated in place with all the new metadata

dlamoris commented on artifacts/src/main/java/org/openmbee/sdvc/artifacts/service/DefaultArtifactService.java:
```diff
@@ -0,0 +1,211 @@
+            ArtifactJson.setArtifacts(elementJson, artifacts);
+            ElementsRequest elementsRequest = new ElementsRequest();
+            elementsRequest.setElements(Arrays.asList(elementJson));
+            nodeService.createOrUpdate(projectId, refId, elementsRequest, params, user);
```
just return from nodeService

dlamoris commented on artifacts/src/main/java/org/openmbee/sdvc/artifacts/json/ArtifactJson.java:
```diff
@@ -0,0 +1,95 @@
+
+        List<Object> rawArtifacts = (List)elementJson.get(ARTIFACTS);
+        if(rawArtifacts == null || rawArtifacts.isEmpty()) {
+            return null;
```
prefer return empty list instead of null, so caller doesn't have to check for nulls

dlamoris commented on json/src/main/java/org/openmbee/sdvc/json/ElementJson.java:
```diff
@@ -4,6 +4,8 @@
 import io.swagger.v3.oas.annotations.media.Schema.AccessMode;
 import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
```
remove unused

BA-MBSE committed:
Transitioned toward throwing exceptions in DefaultArtifactService instead of passing around nulls

BA-MBSE committed:
Changed from using an interceptor to using content negotiation for desired elements endpoint behavior

dlamoris commented on crud/src/main/java/org/openmbee/sdvc/crud/services/DefaultNodeService.java:
```diff
@@ -130,6 +137,8 @@ public ElementsResponse createOrUpdate(String projectId, String refId, ElementsR
         ContextHolder.setContext(projectId, refId);
         boolean overwriteJson = Boolean.parseBoolean(params.get("overwrite"));
 
+        preCommitChanges(projectId, refId, req.getElements(), params, user);
```
Since the artifact subscriber is relying on checking for the artifact controller, this feels like it should really be called at the controller level - it'll basically be a hook for processing input going into /elements - the criteria for applying the processing is url based. The service interface also isn't guaranteed to be invoked through a request handler.

It may be worthwhile to also have a hook for dealing with input before it gets processed at the service level, but I think that would be something along the lines of a subclass override.

BA-MBSE replied:
Reworked this into a generic EmbeddedHookService, and installed the hook at the controller level.

BA-MBSE committed:
Reworked Precommit subscriber service into EmbeddedHookService, which will allow generic hooks to be embedded wherever desired.

BA-MBSE committed:
Cleanup

dlamoris commented on core/src/main/java/org/openmbee/sdvc/core/utils/RequestUtils.java:
```diff
@@ -0,0 +1,52 @@
+package org.openmbee.sdvc.core.utils;
```
can remove since no longer needed

BA-MBSE committed:
Merge branch 'develop' into feature/MMS-98

# Conflicts:
#	core/src/main/java/org/openmbee/sdvc/core/config/Rest.java
#	core/src/main/java/org/openmbee/sdvc/core/config/Utils.java
#	twc/src/main/java/org/openmbee/sdvc/twc/config/Rest.java
